explorer/interpreter/type_checker.cpp

```diff
@@ -1127,6 +1159,10 @@ class TypeChecker::ConstraintTypeBuilder {
     return rewrite_constraints_;
   }
 
+  auto currently_rewriting() const -> std::optional<Nonnull<RewriteInfo*>> {
```

Comment:
I wonder, would it be better if currently_rewriting_ were std::optional<RewriteInfo>? This could return a mutable reference to currently_rewriting_ instead of a pointer (and this function couldn't be const, but it doesn't seem like it's really const if it's allowing state mutation?)